fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/protocol/ProtocolSupport.java

```diff
@@ -0,0 +1,148 @@
+// Copyright 2020 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.metric.protocol;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * An object to describe which protocols a metric can support and to invoke the apply for that protocol.
+ *
+ * In the case where a stack of metrics might be delegated to, white or black listing allows the handler to indicate
+ * whether delegation should be allowed.
+ */
+public class ProtocolSupport {
```

Comment:
I think some wires are getting crossed in this class. My understanding of the Protocol and ProtocolMetric contracts is that multiple different Protocols can share a contract but have different transformers and/or different core parameters. This contract is also intended to support a single ProtocolMetric supporting multiple different Protocols for the same contract (the different types of reagg opperations we have been discussing).

This class implicitly implies that a single metric will only support 1 Protocol per contract name. This is enforced through the instantiation of the `protocolMap` field in the constructor, as well as the accepts, withoutProtocol, withoutProtocols, and getProtocol being based solely on contract name. 

I think this class needs to be refactored to support operations based on solely protocol contract and operations based on the combination of contract + core parameter.

I also now believe that we protocol contract name and protocol core parameter name should be made into concrete classes that wrap strings, which should hopefully clarify some of the contracts here and make these types of errors less likely.

Reply:
I don't agree. It seems to me that a given metric will only have *one* ProtocolHandler, but it does support multiple Protocols. However, we don't support protocol overloading (so a single metric
can't have "currency(String denomination)" and currency(int currencyCode)" on a single metric). We don't want to support protocol overloading. Protocol overloading would be complicated to implement. Making people
use names that are unique for a given metric's namespace is fine.

Reply:
From the discussion Michael and I had it sounds like we want to support protocol contract overloading. The prime example here is with the day of typical week calculation. The current plan is to expose this as a trend metric, while still supporting standard trend metrics. day of typical week and (for example) week over week metrics will use the same contract, but will have different core parameters.

If this design is kept, then ProtocolSupport MUST be keyed by the combination of contract and core parameter. Otherwise, we need to formally change the Protocol contract.

Reply:
ProtocolSupport is immutable on a given metric.  So on a given metric there is exactly one Protocol instance servicing a given contract.

Contract implementations /should/ be the same systemwide.  e.g. it would be weird to support two different protocol implementations with the same contract name on a given system.  Some applications may have different implementations for a given contract than the base implementation.  (e.g. reAgg supporting custom reaggregation types). 

A given protocol has exactly one 'core' parameter which is the signal used to trigger that contract.  It would be sensible, in general, to have the contract name /be/ the coreParameter name, but I can think of edge case exceptions.

Day over Typical Week is a transform triggered that essentially reaggs and trends.  The transform associated with it needs to be attached to either the reagg core parameter or the trend core parameter.  To be applied, it needs to ensure that both the reagg contract and the trend contract are available on the metric it's being applied to.

I was imagining it being implemented on an if-else in the trend protocol.  Essentially from the system point of view it would /be/ a trend, like DayOverDay, just a trend that's illegal if the reagg contract isn't available.

Reply:
I hate making boxes around strings if there isn't some typesafety benefit.